cargo-fix/src/main.rs

```diff
@@ -1,418 +1,191 @@
-                }
-                "r" => {
-                    if suggestion.solutions.len() == 1 && suggestion.solutions[0].replacements.len() == 1 {
-                        let mut solutions = suggestion.solutions;
-                        accepted_suggestions.push(solutions.remove(0).replacements.remove(0));
-                        println!("Suggestion accepted. I'll remember that and apply it later.");
-                        continue 'suggestions;
-                    } else {
-                        println!("{error}: multiple suggestions apply, please pick a number",
-                            error = "Error".red().bold());
-                    }
-                }
-                s => {
-                    if let Ok(i) = usize::from_str(s) {
-                        let replacement = suggestion.solutions
-                            .iter()
-                            .flat_map(|sol| sol.replacements.iter())
-                            .nth(i);
-                        if let Some(replacement) = replacement {
-                            accepted_suggestions.push(replacement.clone());
-                            println!("Suggestion accepted. I'll remember that and apply it later.");
-                            continue 'suggestions;
-                        } else {
-                            println!("{error}: {i} is not a valid suggestion index",
-                                error = "Error".red().bold(),
-                                i = i);
-                        }
-                    } else {
-                        println!("{error}: I didn't quite get that. {user_options}",
-                                    error = "Error".red().bold(),
-                                    user_options = USER_OPTIONS);
-                        continue 'userinput;
-                    }
-                }
-            }
-        }
+        file_map.entry(file_name)
+            .or_insert_with(Vec::new)
+            .push(suggestion);
     }
 
-    if !accepted_suggestions.is_empty() {
-        println!("Good work. Let me just apply these {} changes!",
-                 accepted_suggestions.len());
-
-        for suggestion in accepted_suggestions.iter().rev() {
-            try!(apply_suggestion(suggestion));
-
-            print!(".");
-            flush!();
+    for (file, suggestions) in file_map {
+        // Attempt to read the source code for this file. If this fails then
+        // that'd be pretty surprising, so log a message and otherwise keep
+        // going.
+        let mut code = String::new();
+        if let Err(e) = File::open(&file).and_then(|mut f| f.read_to_string(&mut code)) {
+            warn!("failed to read `{}`: {}", file, e);
+            continue
         }
-
-        println!("\nDone.");
+        let new_code = rustfix::apply_suggestions(&code, &suggestions);
+        File::create(&file)
+            .and_then(|mut f| f.write_all(new_code.as_bytes()))
```

Comment:
Oh, do we need to truncate the file or does create + write_all do that automatically?

Reply:
Ah `File::create` automatically truncates a file that it opens, it's only `OpenOptions::new().write(true)` that won't do that.